### PR TITLE
Dafny 4: User guide updates

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: docs/DafnyRef/VSCodeIDE.md [hunk 1]
```diff
@@ -1,17 +1,33 @@
 # 14. Dafny VSCode extension and the Dafny Language Server {#sec-dafny-language-server-vscode}
 
+## 14.1. Dafny functionality within VSCode
+
 There is a language server for Dafny, which [implements](https://github.com/dafny-lang/dafny/tree/master/Source/DafnyLanguageServer) the [Language Server Protocol](https://microsoft.github.io/language-server-protocol/).
-This server is used by the Dafny VSCode Extension, and it currently offers the following features:
+This server is used by the Dafny VSCode Extension; it currently offers the following features:
 - Quick syntax highlighting
-- Parsing, Resolution and realtime verification diagnostics
+- As you type parsing, resolution and verification diagnostics
```
Comment (alex-chew) [inline docs/DafnyRef/VSCodeIDE.md:8]:
```suggestion
- As-you-type parsing, resolution and verification diagnostics
```

File: docs/DafnyRef/UserGuide.md [hunk 2]
```diff
@@ -31,21 +31,22 @@ are described on the Dafny wiki:
 They are not repeated here to avoid replicating information that
 easily becomes inconsistent and out of date.
 The dafny tool can also be installed using `dotnet tool install --global dafny`
-(given that `dotnet` is already installed on your system).
+(presuming that `dotnet` is already installed on your system).
 
-Most users will find it most convenient to install the pre-built Dafny binaries available on the project release site.
+Most users will find it most convenient to install the pre-built Dafny binaries available on the project release site or in dotnet.
```
Comment (alex-chew) [inline docs/DafnyRef/UserGuide.md:36]:
```suggestion
Most users will find it most convenient to install the pre-built Dafny binaries available on the project release site or using the `dotnet` CLI.
```

File: docs/DafnyRef/UserGuide.md [hunk 3]
```diff
@@ -31,21 +31,22 @@ are described on the Dafny wiki:
...
 As is typical for Open Source projects, dafny can also be built directly from the source files maintained in the github project.
 
 Current and past Dafny binary releases can be found at
 [https://github.com/dafny-lang/dafny/releases](https://github.com/dafny-lang/dafny/releases) for each supported platform.
 Each release is a .zip file with a name combining the release name and the
-platform. Current platforms are Windows 10, Ubuntu 16ff, and MacOS 10.14ff.
+platform. Current platforms are Windows 11, Ubuntu 20ff, and MacOS 10.14ff.
```
Comment (alex-chew) [inline docs/DafnyRef/UserGuide.md:42]: What do the "ff" suffixes mean?

File: docs/DafnyRef/UserGuide.md [hunk 4]
```diff
@@ -249,7 +252,7 @@ Various options control the verification process, in addition to all those descr
 #### 13.5.1.4. `dafny translate <language>` {#sec-dafny-translate}
 
 The `dafny translate` command translates Dafny source code to source code for another target programming language.
-The command always performs the actions of `dafny resolve` and by default does the actions of `dafny verify`.
+The command always performs the actions of `dafny resolve` and, unless `--no-verify is an option`, does the actions of `dafny verify`.
```
Comment (alex-chew) [inline docs/DafnyRef/UserGuide.md:255]:
```suggestion
The command always performs the actions of `dafny resolve` and, unless the `--no-verify` option is specified, does the actions of `dafny verify`.
```

File: docs/DafnyRef/UserGuide.md [hunk 5]
```diff
@@ -361,9 +368,11 @@ The command emits exit codes of
 - 2 for parsing, type-checking or serious errors in running the auditor (e.g. failure to write a report or when report comparison fails)
 - 0 for normal operation, including operation that identifies audit findings
 
+It also takes the `--verbose` option, which then gives information about the files being formatted.
+
 #### 13.5.1.9. `dafny format` {#sec-dafny-format}
 
-Dafny supports a formatter, which for now only changes the indentation of every line in a Dafny file, so that it conforms
+Dafny supports a formatter, which for now only changes the indentation lines in a Dafny file, so that it conforms
```
Comment (alex-chew) [inline docs/DafnyRef/UserGuide.md:375]: I prefer the original wording but it's not a blocker. Could we write "indentation of lines in..."?

File: docs/DafnyRef/UserGuide.md [hunk 6]
```diff
@@ -470,7 +485,7 @@ As is typical for command-line tools, `dafny` provides in-tool help through the
 
 ### 13.5.3. dafny exit codes {#sec-exit-codes}
 
-The dafny tool terminates with these exit codes:
+The basic dafny tool (resolve, verify, translate, build, run) terminates with these exit codes.
```
Comment (alex-chew) [inline docs/DafnyRef/UserGuide.md:488]: What does "basic" mean here?

File: docs/DafnyRef/UserGuide.md [hunk 7]
```diff
@@ -1441,6 +1466,21 @@ These options instruct Dafny to print various information about your
...
+not tied to the local environment (which is a help in testing)
+
+* `--output`, `-o` - location of output files [translate, build]
+
+* `--show-snippets` - include with an error message some of the source code text
+in the neighborhood of the error; the error location (file, line, column) is always given
+
+* `--solver-log <file>` - [verification only] the file in which to place the SMT text sent to the solver
+
+* `--log-format <configuration` - [verification only] logs information about verification performance
```
Comment (alex-chew) [inline docs/DafnyRef/UserGuide.md:1480]:
```suggestion
* `--log-format <configuration>` - [verification only] logs information about verification performance
```

Also, we should list the permitted `configuration`s.

File: docs/DafnyRef/UserGuide.md [hunk 8]
```diff
@@ -1545,6 +1585,23 @@ These options allow some Dafny language features to be enabled or
...
+* `--quantifier-syntax` (value '3' or '4') - permits a choice between the Dafny 3 and Dafny 4 syntax for quantifiers
+
+* `--unicode-char` - if false, the `char` type represents any UTF-16 code unit,
+  that is, any 16-bit value, including surrogate code points and
+  allows `\uXXXX` escapes in string and character literals.
+  If true, `char` represnts any Unicode scalar value,
+  that is, any Unicode code point excluding surrogates and
+  allows `\U{X..X}` escapes in string and character literals. 
+  The default is false for Dafny version 3 and true for version 4.
+  The legacy option was -unicodeChar:<n>` with values 0 and 1 for
```
Comment (alex-chew) [inline docs/DafnyRef/UserGuide.md:1600]:
```suggestion
  The legacy option was `-unicodeChar:<n>` with values 0 and 1 for
```

File: docs/DafnyRef/UserGuide.md [hunk 9]
```diff
@@ -1625,24 +1682,20 @@ older versions of Dafny.
...
-  The default is false for Dafny version 3 and true for version 4.
-  The legacy option was -unicodeChar:<n>` with values 0 and 1 for
-  false and true above.
-
 ### 13.8.6. Controlling warnings {#sec-controlling-warnings}
 
 These options control what warnings Dafny produces, and whether to treat
 warnings as errors.
 
+* `--warn-as-errors (was `-warningsAsErrors`) - treat warnings as errors.
```
Comment (alex-chew) [inline docs/DafnyRef/UserGuide.md:1690]:
```suggestion
* `--warn-as-errors` (was `-warningsAsErrors`) - treat warnings as errors.
```

File: docs/DafnyRef/UserGuide.md [hunk 10]
```diff
@@ -1625,24 +1682,20 @@ older versions of Dafny.
...
+
 * `--warn-shadowing (was `-warnShadowing`) - emit a warning if the name
```
Comment (alex-chew) [inline docs/DafnyRef/UserGuide.md:1692]:
```suggestion
* `--warn-shadowing` (was `-warnShadowing`) - emit a warning if the name 
```

File: docs/DafnyRef/UserGuide.md [hunk 11]
```diff
@@ -2140,3 +2096,128 @@ periods.
...
+  timing, resource use, and outcome for each procedure incrementally, as
+  verification finishes.
+
+* `-randomSeed:<n>` - turn on randomization of the input that Boogie
+  passes to the SMT solver and turn on randomization in the SMT solver
+  itself.
+
+  Certain Boogie inputs are unstable in the sense that changes to the
+  input that preserve its meaning may cause the output to change. The
+  `-randomSeed`` option simulates meaning-preserving changes to the
```
Comment (alex-chew) [inline docs/DafnyRef/UserGuide.md:2132]:
```suggestion
  `-randomSeed` option simulates meaning-preserving changes to the
```

**Review by alex-chew — CHANGES_REQUESTED** [review]

File: docs/DafnyRef/UserGuide.md (hunk 5 above)
Comment (davidcok) [reply docs/DafnyRef/UserGuide.md:375]: That is what I meant. Mikael's words say that every line is altered, which is not the case.

File: docs/DafnyRef/UserGuide.md (hunk 3 above)
Comment (davidcok) [reply docs/DafnyRef/UserGuide.md:42]: And following. But I revised it to be clearer. Besides ff has lots more meanings today. (friends forever?)

File: docs/DafnyRef/UserGuide.md (hunk 6 above)
Comment (davidcok) [reply docs/DafnyRef/UserGuide.md]: Revised

File: docs/DafnyRef/UserGuide.md [hunk 12]
```diff
@@ -1441,6 +1466,21 @@ These options instruct Dafny to print various information about your
...
+not tied to the local environment (which is a help in testing)
+
+* `--output`, `-o` - location of output files [translate, build]
+
+* `--show-snippets` - include with an error message some of the source code text
+in the neighborhood of the error; the error location (file, line, column) is always given
+
+* `--solver-log <file>` - [verification only] the file in which to place the SMT text sent to the solver
+
+* `--log-format <configuration>` - [verification only] logs information about verification performance
```
Comment (alex-chew) [inline docs/DafnyRef/UserGuide.md:1480]: Can we move the explanation of the various permitted configurations / file formats here (from the legacy options section)?

**Review by alex-chew — APPROVED** [review]
Other than the `--log-format` comment above (not a blocker), LGTM.

File: docs/DafnyRef/UserGuide.md (hunk 12 above)
Comment (davidcok) [reply docs/DafnyRef/UserGuide.md:1480]: I'll change this (as I did for other options) in a next PR.

**Pull request merged**